src/librustc/middle/liveness.rs

```diff
@@ -567,7 +552,14 @@ struct Liveness<'a, 'tcx: 'a> {
     tables: &'a ty::TypeckTables<'tcx>,
     s: Specials,
     successors: Vec<LiveNode>,
-    users: Vec<Users>,
+
+    // We used to have a single `users: Vec<Users>` field here, where `Users`
+    // had `reader`, `writer` and `used` fields. But the number of users can
+    // get very large, and it's more compact to store the data in three
+    // separate `Vec`s so that no space is wasted for padding.
+    users_reader: Vec<LiveNode>,
+    users_writer: Vec<LiveNode>,
+    users_used: Vec<bool>,
```

Comment:
Is it worth also making this a bitset/bitvec?

Reply:
https://github.com/rust-lang/rust/issues/54208#issue-360103955 discusses this. It would reduce memory more but increase instruction counts. `Vec<bool>` might be the best middle ground; let's see how the perf results look with it.